### PR TITLE
sequencer: Use batch API to initialize compact.Tree

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change refactors sequencer so that it uses batch API of tree storage to fetch all nodes necessary to construct `compact.Tree`. This should reduce overhead of each sequencing run.

Addresses #1188

### Checklist

- [x] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

File: log/sequencer.go [hunk 1]
```diff
@@ -130,26 +130,30 @@ func NewSequencer(
...
-		if nodes == nil || len(nodes) != 1 {
-			return nil, fmt.Errorf("%x: Did not retrieve one node while loading compact Merkle tree, got %#v for ID %v@%v", s.signer.KeyHint, nodes, nodeID.String(), root.Revision)
+		if got, want := len(nodes), len(storIDs); got != want {
+			return nil, fmt.Errorf("%x: failed to get %d nodes at rev %d, got %d", s.signer.KeyHint, want, root.Revision, got)
 		}
 
-		return nodes[0].Hash, nil
+		hashes := make([][]byte, len(nodes))
+		for i, node := range nodes {
+			hashes[i] = node.Hash
```
Comment (AlCutter) [inline log/sequencer.go]: Given the tightening of the `GetMerkleNodes` contract, I think it's probably worth asserting the assumption holds here - the result of it not holding would be catastrophic...

wdyt?

**Comment by pavelkalinnikov** [comment]
There is a pre-existing lack of coverage for `sequencer.go`. I will add some tests in this PR. Meanwhile, PTAL at the implementation.

File: log/sequencer.go (hunk 1 above)
Comment (pavelkalinnikov) [reply log/sequencer.go]: Great idea. I think I will sleep better if I do this :)

Comment (pavelkalinnikov) [reply log/sequencer.go]: Done.

File: CHANGELOG.md [hunk 2]
```diff
@@ -179,6 +179,10 @@ The `SubtreeCache.GetNodeHash()` method is no longer exported.
 The memory storage provider has been refactored to make it more consistent with
 the other storage providers.
 
+`NodeReader.GetMerkleNodes` now must return `Node` objects in the same order as
```
Comment (Martin2112) [inline CHANGELOG.md]: This was the original intention but the documentation may have accidentally fallen into the sea ...

File: merkle/compact/tree.go [hunk 3]
```diff
@@ -63,27 +69,24 @@ func isPerfectTree(size int64) bool {
...
-
-// getNodeFn will be called a number of times with the coordinates of internal
-// MerkleTree nodes whose hash values are required to initialize the internal
-// state of the compact Tree.  The expectedRoot is the known-good tree root of
-// the tree at size, and is used to verify the correct initial state of the
-// compact Tree after initialisation.
 //
-// TODO(pavelkalinnikov): Make GetNodeFunc get all nodes at once.
-func NewTreeWithState(hasher hashers.LogHasher, size int64, getNodeFn GetNodeFunc, expectedRoot []byte) (*Tree, error) {
+// getNodesFn will be called a number of times with the coordinates of internal
```
Comment (Martin2112) [inline merkle/compact/tree.go]: Isn't this now called once for all nodes?

File: CHANGELOG.md (hunk 2 above)
Comment (pavelkalinnikov) [reply CHANGELOG.md]: Yeah, this paragraph is probably redundant. Are you suggesting to drop it?

Comment (Martin2112) [reply CHANGELOG.md]: I think I'd leave it in as I don't think we've documented it properly before.

File: merkle/compact/tree.go (hunk 3 above)
Comment (pavelkalinnikov) [reply merkle/compact/tree.go]: Indeed. Updated the comment.

**Comment by Martin2112** [comment]
FYI I don't think this overlaps with my experiment in #1535 but I'm expecting it to speed up compact range hashing.

**Comment by pavelkalinnikov** [comment]
Added tests in #1543, will rebase this one for better coverage.

**Pull request merged**